pkg/lib/config/configsource/database.go

```diff
@@ -34,22 +40,54 @@ func NewDatabaseLogger(lf *log.Factory) DatabaseLogger {
 }
 
 type Database struct {
-	Logger        DatabaseLogger
-	BaseResources *resource.Manager
-	TrustProxy    config.TrustProxy
-	Config        *Config
-	Clock         clock.Clock
+	Logger         DatabaseLogger
+	BaseResources  *resource.Manager
+	TrustProxy     config.TrustProxy
+	Config         *Config
+	Clock          clock.Clock
+	Store          *Store
+	Database       *globaldb.Handle
+	DatabaseConfig *config.DatabaseEnvironmentConfig
 
 	done     chan<- struct{} `wire:"-"`
-	Store    *Store
-	Database *globaldb.Handle
+	listener *db.PQListener  `wire:"-"`
+
+	hostMap *sync.Map `wire:"-"`
+	appMap  *sync.Map `wire:"-"`
 }
 
 func (d *Database) Open() error {
+	d.hostMap = &sync.Map{}
+	d.appMap = &sync.Map{}
+
+	done := make(chan struct{})
+	d.done = done
+
+	d.listener = &db.PQListener{
+		DatabaseURL: d.DatabaseConfig.DatabaseURL,
+	}
+	go d.listener.Listen([]string{
+		PGChannelConfigSourceChange,
+		PGChannelDomainChange,
+	}, done, func(channel string, extra string) {
+		switch channel {
+		case PGChannelConfigSourceChange:
+			d.invalidateApp(extra)
+		case PGChannelDomainChange:
+			d.invalidateHost(extra)
+		default:
+			panic(fmt.Sprintf("db_config: unknown notification channel: %s", channel))
+		}
+	}, func(e error) {
+		panic(fmt.Sprintf("db_config: error on listening pgsql: %s", e))
```

Comment:
If we panic here, this goroutine will die. However, this config source will no longer listen anything :(

Reply:
Tested the listener has mechanism to reconnect, so we don't need to call `panic` here to break the program. Changed to log the error instead.